linked_list/linked_list.py

```diff
@@ -1,75 +1,170 @@
+                    return True
+                else:
+                    return False
+            elif current.value == value:
+                return True
+            else:
+                current = current.next
+        return False
+    
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):
-        pass
+        counter = 0
+
+        if self.head == None:
+            return 0
+        current = self.head
+        while current is not None:
+            counter += 1
+            current = current.next 
+        return counter
+
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(self, index):
-        pass
+        counter = 0
+        
+        if self.head == None:
+            return None
+
+        current = self.head
+        while current is not None:
+            if counter == index:
+                return current.value
+            else:
+                current = current.next
+                counter += 1
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        current = self.head
+
+        while current:
+            if current.next is None:
+                return current.value
+            else:
+                current = current.next
+
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        if self.tail is None:
+            self.tail = self.head = Node(value)
+        else:
+            new_node = Node(value, None, self.tail)
+            self.tail.next = new_node
+            self.tail = new_node
+
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_max(self):
-        pass
+        max_val = 0
+
+        if self.head == None:
+            max_val = None
+        current = self.head
+        while current is not None:
+            if current.value > max_val:
+                max_val = current.value
+            current = current.next
+        return max_val
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def delete(self, value):

```

Comment:
👍  Just note that you can notice here where a doubly linked list is a bit easier to write for this method.